incremental.go

```diff
@@ -0,0 +1,267 @@
+	p := &IncrementalProducer{
+		bucket: bucket,
+		ctx:    ctx,
+		stop:   stop,
+		opt:    o,
+		modfn:  modfn,
+		pfn:    pfn,
+	}
+
+	// run initial push
+	if _, err := p.push(); err != nil {
+		_ = p.Close()
+		return nil, err
+	}
+
+	// start continuous loop
+	// use error group
+	go p.loop()
+
+	return p, nil
+}
+
+// Close stops the producer.
+func (p *IncrementalProducer) Close() error {
+	p.stop()
+	if p.ownBucket {
+		return p.bucket.Close()
+	}
+	return nil
+}
+
+// LastPush returns time of last push attempt.
+func (p *IncrementalProducer) LastPush() time.Time {
+	return timestamp(atomic.LoadInt64(&p.lastPush)).Time()
+}
+
+// LastModified returns time at which the remote feed was last modified.
+func (p *IncrementalProducer) LastModified() time.Time {
+	return timestamp(atomic.LoadInt64(&p.lastMod)).Time()
+}
+
+// NumWritten returns the number of values produced during the last push.
+func (p *IncrementalProducer) NumWritten() int {
+	return int(atomic.LoadInt64(&p.numWritten))
+}
+
+func (p *IncrementalProducer) loadManifest() (*Manifest, error) {
+	remote := bfs.NewObjectFromBucket(p.bucket, "manifest.json")
+	defer remote.Close()
+
+	m := new(Manifest)
+
+	r, err := NewReader(p.ctx, remote, nil)
+	if errors.Is(err, bfs.ErrNotFound) {
+		return m, nil
+	} else if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	if err := r.Decode(m); errors.Is(err, bfs.ErrNotFound) { // some BFS implementations defer Open-ing the S3 object till first Decode call
+		return m, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+func (p *IncrementalProducer) loop() {
+	ticker := time.NewTicker(p.opt.Interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-ticker.C:
+			state, err := p.push()
+			if p.opt.AfterPush != nil {
+				p.opt.AfterPush(state, err)
+			}
+		}
+	}
+}
+
+func (p *IncrementalProducer) push() (*IncrementalProducerPush, error) {
+	start := time.Now()
+	atomic.StoreInt64(&p.lastPush, timestampFromTime(start).Millis())
+
+	// get last mod time for local records
+	localLastMod, err := p.modfn(p.ctx)
+	if err != nil {
+		return nil, err
+	}
+	if localLastMod.IsZero() {
+		return &IncrementalProducerPush{IncrementalProducer: p}, nil
+	}
+
+	// fetch manifest from remote
+	manifest, err := p.loadManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	// compare manifest LastModified to local last mod.
+	remoteLastMod := manifest.LastModified
+	if remoteLastMod == timestampFromTime(localLastMod) {
+		return &IncrementalProducerPush{IncrementalProducer: p}, nil
+	}
+
+	// TODO! should the meta LastMod be the file modified time or the local record last updated time?
+	wopt := p.opt.WriterOptions
+	wopt.LastMod = start
+
+	// write modified data
+	// TODO! set the file extension from WriterOpts format & compression
+	fname := "data-" + strconv.Itoa(manifest.Generation) + "-" + localLastMod.Format("2006-01-02-15:04:05.0000") + ".pbz"
```

Comment:
that's not a good time format for file names, please use 20060102-150405000 or similar. also, careful with file extensions naming, it's not always pbz but depends on format!

Reply:
yup, I need to add a method to get extension from format+compression, its a TODO above